### PR TITLE
Proposal to create language family builders

### PR DESCRIPTION
**Pull request by kevin-ortega**
[Readable](https://github.com/kevin-ortega/rfcs/blob/create-java-specific-builder/text/0000-create-language-family-builders.md)

### CODE REVIEW

**Comment by ryanmoran** [comment]
Today we offer 3 different "flavors" of builder based on the stack they align with. How would we deliver a coherent set of language-specific builders as we offer more stacks (or "flavors") going forward?

For instance, what does it mean to have a builder called `paketobuildpacks/builder:nodejs` or `paketobuildpacks/builder:go`? Would the Node.js builder build on top of the `base` Jammy stack while the Go builder built off the `tiny` stack? And what of other stacks? Where does this leave UBI or a future Alpine stack?

**Comment by kevin-ortega** [comment]
> Today we offer 3 different "flavors" of builder based on the stack they align with. How would we deliver a coherent set of language-specific builders as we offer more stacks (or "flavors") going forward?
> 
> For instance, what does it mean to have a builder called `paketobuildpacks/builder:nodejs` or `paketobuildpacks/builder:go`? Would the Node.js builder build on top of the `base` Jammy stack while the Go builder built off the `tiny` stack? And what of other stacks? Where does this leave UBI or a future Alpine stack?

I'm updating the RFC on how to deal with jammy, ubi and future ubuntu-based stacks.  I'm not familiar enough with the other language families and the full, base, tiny flavors and what each family would need.    Is it sufficient for the RFC to suggest each language family will need to choose what stack their respective builder should be based on?

**Comment by ryanmoran** [comment]
I think I can get onboard with the current language about the variants that would be available for each language family.

I also see that the `builder.toml` example provided composes the component buildpacks that make up the current Java buildpack. Is the expectation that we would get rid of the language-family buildpacks and replace them with builders. My concern with this is that we end up with significant duplication, or even incidental skew in the buildpacks included in each builder. It also prevents someone from easily using a "buildpackless" builder and the language-family buildpack to build.

What are the benefits of pushing the language-family order configurations into the builders?

**Comment by sophiewigmore** [comment]
WG notes:
Looking for feedback from maintainers of other language families and builders

**Comment by kevin-ortega** [comment]
> I think I can get onboard with the current language about the variants that would be available for each language family.
> 
> I also see that the `builder.toml` example provided composes the component buildpacks that make up the current Java buildpack. Is the expectation that we would get rid of the language-family buildpacks and replace them with builders. My concern with this is that we end up with significant duplication, or even incidental skew in the buildpacks included in each builder. It also prevents someone from easily using a "buildpackless" builder and the language-family buildpack to build.
> 
> What are the benefits of pushing the language-family order configurations into the builders?

I don't propose to replace language-family buildpacks with builders.   I will update the sample builder.toml to reflect that.

File: text/0000-create-language-family-builders.md
Comment (dmikusa) [inline text/0000-create-language-family-builders.md:33]: Is this OK with @paketo-buildpacks/builders-maintainers? Is it technically possible to add additional builders this way?

Also, I would kind of like to see this jointly owned. Where the @paketo-buildpacks/builders-maintainers own the machinery/workflows to produce the images, but the language team owns the actual builder.toml & any tests for validation.

Thoughts?

Comment (dmikusa) [reply text/0000-create-language-family-builders.md:33]: I think we should also make a note here to say something like, basically to make it clear that ownership is delegated to the language family teams. RFCs and decisions about these builders would go through the language family RFC process.

```
The decision to create language family-specific builders is owned by each language family team. Teams may decide to create zero or more based on their needs and user demand. In addition, each language-family team owns the decisions about what goes into the builder. Again, this can be based on team and user needs.
```

Comment (kevin-ortega) [reply text/0000-create-language-family-builders.md]: @dmikusa I have updated the RFC to more clearly propose joint ownership and delegate the creation to each individual language family.

Comment (dmikusa) [reply text/0000-create-language-family-builders.md]: Thanks Kevin. I'm going to resolve this conversation. I've added @paketo-buildpacks/builders-maintainers as a reviewer, so we can get some feedback.

**Review by dmikusa — APPROVED** [review]

**Comment by johnnyr0x** [comment]
@ryanmoran to follow up with team to get more input

**Comment by dmikusa** [comment]
@paketo-buildpacks/builders-maintainers - Would you be able to weigh in on this one? Primarily, the ownership is going to be on the language-family team, but language-family teams will use builder tools so there is some shared ownership. Thanks!

**Review by robdimsdale — APPROVED** [review]
This looks good to me.

The only thing I'd call out is that OSS support for Bionic will be dropped in April 2023 - both by Paketo but also by Canonical - so by the time we approve and implement this RFC there may not be much value in implementing a Bionic builder for language families. Therefore we may wish to remove it from the RFC. I don't mind if we keep it in the RFC though in the interests of getting it approved sooner.

The ownership detailed in the Implementation section seems fine, with the addition that some of the machinery for the builders is co-maintained with @paketo-buildpacks/tooling-maintainers (i.e. the [`builder` section of the shared github config](https://github.com/paketo-buildpacks/github-config/tree/main/builder))

**Comment by kevin-ortega** [comment]
> This looks good to me.
> 
> The only thing I'd call out is that OSS support for Bionic will be dropped in April 2023 - both by Paketo but also by Canonical - so by the time we approve and implement this RFC there may not be much value in implementing a Bionic builder for language families. Therefore we may wish to remove it from the RFC. I don't mind if we keep it in the RFC though in the interests of getting it approved sooner.
> 
> The ownership detailed in the Implementation section seems fine, with the addition that some of the machinery for the builders is co-maintained with @paketo-buildpacks/tooling-maintainers (i.e. the [`builder` section of the shared github config](https://github.com/paketo-buildpacks/github-config/tree/main/builder))

Good point.  I have updated the RFC and removed bionic examples.

**Review by ryanmoran — APPROVED** [review]

**Comment by sophiewigmore** [comment]
Working group 3/15 - waiting on a final review from @ForestEckhardt

**Comment by robdimsdale** [comment]
I think it will be easier for the builders team to maintain if we have separate repos per builder. So one repo for `builder-jammy-full-java`, one for `builder-jammy-tiny-java`, etc. The existing tooling should work fairly well with minimal overhead for that structure. If we want to adopt a monorepo approach we'll have to spend additional time and effort adapting the existing automation to work with that. Not an inherent problem - just pointing it out.

**Comment by ForestEckhardt** [comment]
I am fine with separate repos from a builder maintainer perspective but that does increase the number of repo for participating language families by a quite a bit. Would just want to make sure that we are comfortable with adding that level of overhead.

**Comment by robdimsdale** [comment]
Great, so in terms of naming, the current (jammy) naming convention is as follows: `builder-<OS>-<varient>` e.g. `builder-jammy-full`.

For the language-specific subtypes, I'd propose adding the subtype as a suffix (rather than somewhere else like in the middle or as a prefix). So that would look like: `builder-jammy-full-java` or `builder-jammy-base-java-with-extras`.

**Comment by kevin-ortega** [comment]
> Great, so in terms of naming, the current (jammy) naming convention is as follows: `builder-<OS>-<variant>` e.g. `builder-jammy-full`.
> 
> For the language-specific subtypes, I'd propose adding the subtype as a suffix (rather than somewhere else like in the middle or as a prefix). So that would look like: `builder-jammy-full-java` or `builder-jammy-base-java-with-extras`.

I agree and I will update the RFC with this naming convention.

File: text/0000-create-language-family-builders.md
Comment (robdimsdale) [inline text/0000-create-language-family-builders.md:48]: I assume this was meant to read: `builder-jammy-base-java`?

Comment (kevin-ortega) [reply text/0000-create-language-family-builders.md:48]: > I assume this was meant to read: `builder-jammy-base-java`?

Crap, yes.  Thank you. I’ll fix.

**Review by robdimsdale — APPROVED** [review]
LGTM

**Review by ForestEckhardt — APPROVED** [review]

**Comment by dmikusa** [comment]
Thanks for the comments and votes. I think we have the votes to proceed, so this is the final chance to comment.

If there are no additional comments or objections by EOD Friday, Mar 24th, then we'll go ahead and merge this. Thanks!

**Pull request merged**
